### PR TITLE
*: add a API to reset the TS

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve? 
For backup tool. Supports this API.

### What is changed and how it works?
add an API.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: server/tso/tso.go [hunk 1]
```diff
@@ -35,12 +37,13 @@ const (
 	UpdateTimestampStep  = 50 * time.Millisecond
 	updateTimestampGuard = time.Millisecond
 	maxLogical           = int64(1 << 18)
+	maxResetGap          = 24 * time.Hour
```
Comment (overvenus) [inline server/tso/tso.go]: Could you make it configurable?

File: server/tso/tso.go [hunk 2]
```diff
@@ -134,11 +137,41 @@ func (t *TimestampOracle) SyncTimestamp(lease *member.LeaderLease) error {
...
+	next := pythsic.Add(time.Millisecond)
+	prev := (*atomicObject)(atomic.LoadPointer(&t.ts))
+
+	// do not update
+	if typeutil.SubTimeByWallClock(next, prev.physical) <= 3*updateTimestampGuard {
+		return errors.New("cannot be reset the smaller tso")
+	}
+
+	if typeutil.SubTimeByWallClock(next, prev.physical) >= maxResetGap {
+		return errors.New("the specified ts too large than now")
```
Comment (overvenus) [inline server/tso/tso.go]: Could you record the error to `tsoCounter`?

File: server/tso/tso.go [hunk 3]
```diff
@@ -134,11 +137,41 @@ func (t *TimestampOracle) SyncTimestamp(lease *member.LeaderLease) error {
...
+	}
+
+	save := next.Add(t.saveInterval)
+	if err := t.saveTimestamp(save); err != nil {
+		return err
```
Comment (overvenus) [inline server/tso/tso.go]: Could you record the error to `tsoCounter`?

File: server/tso/tso.go [hunk 4]
```diff
@@ -134,11 +137,41 @@ func (t *TimestampOracle) SyncTimestamp(lease *member.LeaderLease) error {
...
+	if t.lease == nil || t.lease.IsExpired() {
+		return errors.New("Setup timestamp failed, lease expired")
+	}
+	pythsic, _ := tsoutil.ParseTS(tso)
+	next := pythsic.Add(time.Millisecond)
+	prev := (*atomicObject)(atomic.LoadPointer(&t.ts))
+
+	// do not update
+	if typeutil.SubTimeByWallClock(next, prev.physical) <= 3*updateTimestampGuard {
+		return errors.New("cannot be reset the smaller tso")
```
Comment (overvenus) [inline server/tso/tso.go]: Could you record the error to `tsoCounter`?

File: server/tso/tso.go [hunk 5]
```diff
@@ -35,12 +37,13 @@ const (
...
 	updateTimestampGuard = time.Millisecond
 	maxLogical           = int64(1 << 18)
+	maxResetGap          = 24 * time.Hour
 )
 
 // TimestampOracle is used to maintain the logic of tso.
 type TimestampOracle struct {
 	// For tso, set after pd becomes leader.
-	ts            atomic.Value
+	ts            unsafe.Pointer
```
Comment (lhy1024) [inline server/tso/tso.go]: Is it necessary to use unsafe?

File: server/tso/tso.go [hunk 6]
```diff
@@ -134,11 +137,41 @@ func (t *TimestampOracle) SyncTimestamp(lease *member.LeaderLease) error {
...
 
 	return nil
 }
 
+// ResetUserTimestamp update the physical part with specified tso.
+func (t *TimestampOracle) ResetUserTimestamp(tso int64) error {
+	if t.lease == nil || t.lease.IsExpired() {
+		return errors.New("Setup timestamp failed, lease expired")
+	}
+	pythsic, _ := tsoutil.ParseTS(tso)
```
Comment (disksing) [inline server/tso/tso.go]: typo

File: server/tso/tso.go (hunk 5 above)
Comment (nolouch) [reply server/tso/tso.go]: We need CAS operator now.

File: server/tso/tso.go [hunk 7]
```diff
@@ -191,6 +234,7 @@ func (t *TimestampOracle) UpdateTimestamp() error {
 	if typeutil.SubTimeByWallClock(t.lastSavedTime, next) <= updateTimestampGuard {
 		save := next.Add(t.saveInterval)
 		if err := t.saveTimestamp(save); err != nil {
+			tsoCounter.WithLabelValues("err_save_ts").Inc()
```
Comment (rleungx) [inline server/tso/tso.go]: ditto, should be different from line 134?

Comment (nolouch) [reply server/tso/tso.go]: they are the same reason.

**Comment by nolouch** [comment]
PTAL @lhy1024 @rleungx @overvenus

File: server/api/router.go [hunk 8]
```diff
@@ -141,6 +141,8 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	router.Handle("/api/v1/health", newHealthHandler(svr, rd)).Methods("GET")
 	router.Handle("/api/v1/diagnose", newDiagnoseHandler(svr, rd)).Methods("GET")
 
+	router.HandleFunc("/api/v1/recovery/tso", newRecoveryHandler(handler, rd).ResetTS).Methods("POST")
```
Comment (shafreeck) [inline server/api/router.go]: `recovery` is just a single feature, it should not occupy the api's name. What if there is a new feature that needs to reset the TSO too?

File: server/api/recovery.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"net/http"
+
+	"github.com/pingcap/pd/pkg/apiutil"
+	"github.com/pingcap/pd/server"
+	"github.com/unrolled/render"
+)
+
+type recoveryHandler struct {
```
Comment (shafreeck) [inline server/api/recovery.go]: I think we should not call `recovery*` here including the filename. Supporting the recovery feature is our target to change the code this time but not for the code self.

We should name it according to what we want to supply in the view of PD.

File: server/config/config.go [hunk 10]
```diff
@@ -831,12 +832,17 @@ func (s SecurityConfig) ToTLSConfig() (*tls.Config, error) {
 type PDServerConfig struct {
 	// UseRegionStorage enables the independent region storage.
 	UseRegionStorage bool `toml:"use-region-storage" json:"use-region-storage,string"`
+	// MaxResetTSGap the max gap to reset the tso
```
Comment (shafreeck) [inline server/config/config.go]:
```suggestion
	// MaxResetTSGap is the max gap to reset the tso
```

File: server/tso/tso.go (hunk 5 above)
Comment (shafreeck) [reply server/tso/tso.go]: Could we use a meaningful pointer name like `*atomic.Value`? Covert it to an `unsafe.Pointer` when it is used.

File: pkg/tsoutil/tso.go [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tsoutil
+
+import "time"
+
+const (
+	physicalShiftBits = 18
+	logicalBits       = 0x3FFFF
```
Comment (disksing) [inline pkg/tsoutil/tso.go]: `(1<< physicalShiftBits)-1`?

File: server/api/recovery.go (hunk 9 above)
Comment (disksing) [reply server/api/recovery.go]: how about `/admin/reset_ts`?

File: server/tso/tso.go [hunk 12]
```diff
@@ -40,22 +43,26 @@ const (
...
+	ts            unsafe.Pointer
 	lastSavedTime time.Time
 	lease         *member.LeaderLease
 
 	rootPath     string
 	member       string
 	client       *clientv3.Client
 	saveInterval time.Duration
+	// TODO: rename ScheduleOption
+	opt *config.ScheduleOption
```
Comment (disksing) [inline server/tso/tso.go]: I think you can define a local interface that includes `GetTSGap()`.

File: server/api/router.go [hunk 13]
```diff
@@ -134,6 +134,7 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 
 	adminHandler := newAdminHandler(svr, rd)
 	router.HandleFunc("/api/v1/admin/cache/region/{id}", adminHandler.HandleDropCacheRegion).Methods("DELETE")
+	router.HandleFunc("/api/v1/admin/reset_ts", adminHandler.ResetTS).Methods("POST")
```
Comment (rleungx) [inline server/api/router.go]:
```suggestion
	router.HandleFunc("/api/v1/admin/reset-ts", adminHandler.ResetTS).Methods("POST")
```

File: server/server.go [hunk 14]
```diff
@@ -221,7 +221,7 @@ func (s *Server) startServer() error {
 	s.member.MemberInfo(s.cfg, s.Name(), s.rootPath)
 
 	s.idAllocator = id.NewAllocatorImpl(s.client, s.rootPath, s.member.MemberValue())
-	s.tso = tso.NewTimestampOracle(s.client, s.rootPath, s.member.MemberValue(), s.cfg.TsoSaveInterval.Duration)
+	s.tso = tso.NewTimestampOracle(s.client, s.rootPath, s.member.MemberValue(), s.cfg.TsoSaveInterval.Duration, func() time.Duration { return s.scheduleOpt.LoadPDServerConfig().MaxResetTSGap })
```
Comment (lhy1024) [inline server/server.go]: would you like to split it into multiple lines

File: server/api/router.go [hunk 15]
```diff
@@ -134,15 +134,14 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 
 	adminHandler := newAdminHandler(svr, rd)
 	router.HandleFunc("/api/v1/admin/cache/region/{id}", adminHandler.HandleDropCacheRegion).Methods("DELETE")
+	router.HandleFunc("/api/v1/admin/reset-ts", adminHandler.ResetTS).Methods("POST")
```
Comment (shafreeck) [inline server/api/router.go]: How about `POST /api/v1/admin/ts`? An action(`reset` here) is not recommended in the URL.

Comment (nolouch) [reply server/api/router.go]: I think reset more meaningful, it's a dangerous action.

**Pull request merged**